### DISCUSSION TITLE
How to use sympy in firedrake

### DISCUSSION THREAD

**Discussion by zhyxue0** [firedrakeproject/firedrake#3187] [Firedrake support]
Hello, everyone! I am trying to use `sympy` in firerake to make expressions general.

Here is my idea, but it seems does not work.
```python
import sympy

x, y, t = sympy.symbols('x y t')
u = 3*x*y + x**2 + 2*y**2 + t**3

```
Do you have any good suggestions?

**Comment by wence-** [comment]
What would you like to do with the expression? It may be that you can get away with using UFL for your symbolics (and the things will be easier).

**Reply by zhyxue0** [reply]
For example, given a exact solution expression u.  I try to solve the heat conduction equation u_t = a(u_xx + u_yy) + f, and obtain the corresponding expression of the source term `f`. `f` depends on the expression of the exact solution u (actually depends on `x,y,t`). I want a way to get a unified expression of `f`, so that `f` can be updated in time iteration.

**Reply by wence-** [reply]
OK, so you want to symbolically differentiate to find `f = u_t - a(u_xx + u_yy)`. You can do this with ufl if you use UFL objects and `variable` and `diff` (see https://fenics.readthedocs.io/projects/ufl/en/latest/manual/form_language.html#variable-derivatives)

So, suppose you have x y and z as being spatial coordinates:

```python
x, y = SpatialCoordinate(mesh)
t = Constant(...) # current time

# define objects that we can differentiate with respect to
xv = variable(x)
yv = variable(y)
tv = variable(t)

u = 3*xv*yv + xv**2 + 2*yv**2 + tv**3
# a is defined previous as a normal python function
f = diff(u, tv) - a(diff(diff(u, xv), xv) + diff(diff(u, yv), yv))

# Now you can just use `f` as a normal UFL expression
```

**Comment by rckirby** [comment]
This is also done several places in Irksome for time-dependent problems.
For example, check the demo:

https://github.com/firedrakeproject/Irksome/blob/master/demos/heat/demo_heat.py.rst

There are other demos and tests in similar vein.

**Reply by zhyxue0** [reply]
Thanks a lot. But I have some questions. I note 
``` python
uexact = B * atan(t)*(pi / 2.0 - atan(S * (R - t)))
rhs = expand_derivatives(diff(uexact, t)) - div(grad(uexact))
```


f I want to solve for `f` by this equation, $u=(u_0, u_1)$ is a vector,a vector, therefore `f` is also a vector,
Does `expand_derivatives(diff(uexact, t)) - div(grad(uexact))` also work in vector form?

And this equation is 
$f = u_{t} -  \Delta u - |u|^{p} u$, where $|u| =（u_0^2 + u_1^2)^p$. 
How to express `f` if I code 
``` python
def u_exact(x, y, t):
    u = as_vector(sin(x)*t, sin(y)*t)
return u


# How to code it?
def f(u):

```

**Comment by rckirby** [comment]
It should respect the rules of vector algebra/calculus.  You can stick UFL
expressions into arrays by `as_vector` (it takes a list?) and then you need
to compute the vector Laplacian that's right for you.  You should try it
out and post specific questions when something breaks.
